apiserver/Makefile

```diff
@@ -43,16 +17,75 @@ jupyter_notebook_config.py:
-$(eval $(call DOCKER_IMAGE,spark-master,install-hail-locally jupyter_notebook_config.py,SPARK_MASTER_IMAGE))
-$(eval $(call DOCKER_IMAGE,spark-worker,install-hail-locally jupyter_notebook_config.py,SPARK_WORKER_IMAGE))
-$(eval $(call DOCKER_IMAGE,apiserver,install-hail-locally jupyter_notebook_config.py,APISERVER_IMAGE))
-$(eval $(call DOCKER_IMAGE,hail-jupyter,install-hail-locally jupyter_notebook_config.py,HAIL_JUPYTER_IMAGE))
+ifeq ($(IN_HAIL_CI),1)
+build-spark-base:
+	docker pull debian:9.5
+	-docker pull gcr.io/$(PROJECT)/spark-base:latest
+	docker build -t spark-base -f Dockerfile.spark-base . --cache-from spark-base,gcr.io/$(PROJECT)/spark-base:latest,debian:9.5
+
+build-hail-base: build-spark-base install-hail-locally
+	-docker pull gcr.io/$(PROJECT)/spark-base:latest
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	docker build -t hail-base -f Dockerfile.hail-base . --cache-from hail-base,gcr.io/$(PROJECT)/hail-base:latest,spark-base
+
+build-spark-master: build-hail-base
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/spark-master:latest
+	docker build -t spark-master -f Dockerfile.spark-master . --cache-from spark-master,gcr.io/$(PROJECT)/spark-master:latest,hail-base
+
+build-spark-worker: build-hail-base
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/spark-worker:latest
+	docker build -t spark-worker -f Dockerfile.spark-worker . --cache-from spark-worker,gcr.io/$(PROJECT)/spark-worker:latest,hail-base
+
+build-apiserver: build-hail-base
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/apiserver:latest
+	docker build -t apiserver -f Dockerfile.apiserver . --cache-from apiserver,gcr.io/$(PROJECT)/apiserver:latest,hail-base
+
+build-hail-jupyter: build-hail-base jupyter_notebook_config.py
+	-docker pull gcr.io/$(PROJECT)/hail-base:latest
+	-docker pull gcr.io/$(PROJECT)/hail-jupyter:latest
+	docker build -t hail-jupyter -f Dockerfile.hail-jupyter . --cache-from hail-jupyter,gcr.io/$(PROJECT)/hail-jupyter:latest,hail-base
+
+build: build-spark-master build-spark-worker build-apiserver build-hail-jupyter
+
+SPARK_MASTER_IMAGE = "gcr.io/$(PROJECT)/spark-master:$(shell docker images -q --no-trunc spark-master | sed -e 's,[^:]*:,,')"
+SPARK_WORKER_IMAGE = "gcr.io/$(PROJECT)/spark-worker:$(shell docker images -q --no-trunc spark-worker | sed -e 's,[^:]*:,,')"
+APISERVER_IMAGE = "gcr.io/$(PROJECT)/apiserver:$(shell docker images -q --no-trunc apiserver | sed -e 's,[^:]*:,,')"
+HAIL_JUPYTER_IMAGE = "gcr.io/$(PROJECT)/hail-jupyter:$(shell docker images -q --no-trunc hail-jupyter | sed -e 's,[^:]*:,,')"
+
+tag: build
+	docker tag spark-master $(SPARK_MASTER_IMAGE)
+	docker tag spark-worker $(SPARK_WORKER_IMAGE)
+	docker tag apiserver $(APISERVER_IMAGE)
+	docker tag hail-jupyter $(HAIL_JUPYTER_IMAGE)
+
+push: tag
+	docker push $(SPARK_MASTER_IMAGE)
```

Comment:
As @danking said, you need to push the latest tag also, for caching.

Reply:
Ah, thanks. That's strange friction to introduce. "latest" should be a pointer to the last image pushed.

Reply:
Latest should be called something like default.  It doesn't mean what you think it means: https://vsupalov.com/docker-latest-tag/

Reply:
Yeah, I just read that and another source. Sbt (have not used, came up in search on the latest tag) provides an option to force auto-update of latest tag, which is nice: https://www.scala-sbt.org/sbt-native-packager/formats/docker.html